File: ELEC881_Tabnine/src/No Description Prompt/Java script (JS)/q19-can-i-win.js
/**
 * @param {number} maxChoosableInteger
 * @param {number} desiredTotal
 * @return {boolean}
 */
var canIWin = function(maxChoosableInteger, desiredTotal) {
    if (desiredTotal <= 0) return true;
    if (maxChoosableInteger * (maxChoosableInteger + 1) / 2 < desiredTotal) return false;

    const dp = new Array(desiredTotal + 1).fill(false);
    dp[0] = true;

    for (let i = 1; i <= desiredTotal; i++) {
        for (let j = 1; j <= maxChoosableInteger; j++) {
            if (i - j >= 0 &&!dp[i - j]) {
                dp[i] = true;
                break;
            }
        }
    }

    return dp[desiredTotal];
};

File: ELEC881_Copilot/src/No Description Prompt/Python/q19-can-i-win.py
class Solution(object):
    def canIWin(self, maxChoosableInteger, desiredTotal):
        """
        :type maxChoosableInteger: int
        :type desiredTotal: int
        :rtype: bool
        """
        if maxChoosableInteger >= desiredTotal:
            return True
        if (1 + maxChoosableInteger) * maxChoosableInteger / 2 < desiredTotal:
            return False
        return self.canIWinHelper(maxChoosableInteger, desiredTotal, 0, {}) == 1
    
    def canIWinHelper(self, maxChoosableInteger, desiredTotal, state, memo):
        if state in memo:
            return memo[state]
        for i in range(maxChoosableInteger):
            if (state & (1 << i)) == 0:
                if i + 1 >= desiredTotal or not self.canIWinHelper(maxChoosableInteger, desiredTotal - i - 1, state | (1 << i), memo):
                    memo[state] = 1
                    return 1
        memo[state] = 0
        return 0